File: packages/ruby/lib/opensms/resources/suppressions.rb
# frozen_string_literal: true

require_relative "base"

module Opensms
  module Resources
    # The +suppressions+ resource: numbers that must not receive messages.
    # Accessed as +client.suppressions+.
    class Suppressions < Base
      # GET /v1/compliance/suppressions -> Page<Suppression>.
      def list(params = nil, **kwargs)
        http_page("/v1/compliance/suppressions", Models.build(merge(params, kwargs), %i[limit cursor]))
      end

      # POST /v1/compliance/suppressions -> 201 Suppression. Not auto-retried.
      # @param params [Hash] :e164, :reason (stop_keyword|manual|complaint|invalid_number)
      def create(params = nil, **kwargs)
        body = Models.build(merge(params, kwargs), %i[e164 reason], required: %i[e164 reason])
        @transport.request(:post, "/v1/compliance/suppressions", body: body)
      end

      # Bulk add. POST /v1/compliance/suppressions/import -> 201 { created:, received: }.
      # Not auto-retried.
      #
      # @param items [Array<Hash>] each { e164:, reason: }
      def import(items)
        raise ArgumentError, "Opensms: `items` must be an Array." unless items.is_a?(Array)

        body = { items: items.map { |i| Models.build(i, %i[e164 reason], required: %i[e164 reason]) } }
        @transport.request(:post, "/v1/compliance/suppressions/import", body: body)
      end

      # DELETE /v1/compliance/suppressions/{id} -> 204.
      # @return [nil]
      def delete(id)
        http_delete(path("/v1/compliance/suppressions", id!(id)))
      end
    end
  end
end

File: packages/python/src/opensms/resources/suppressions.py
"""The ``suppressions`` resource."""

from __future__ import annotations

from typing import Iterable, Mapping, Optional, Union

from .._serialize import path_id
from ..models import Page, Suppression, SuppressionImportResult
from ._base import Resource


class Suppressions(Resource):
    """Numbers that must never be messaged. Accessed as ``client.suppressions``."""

    def list(self, *, limit: Optional[int] = None, cursor: Optional[str] = None) -> Page[Suppression]:
        """List suppressions."""
        return self._page(
            self._t.request("GET", "/v1/compliance/suppressions", query={"limit": limit, "cursor": cursor})
        )

    def create(self, *, e164: str, reason: str) -> Suppression:
        """Suppress a number (``reason`` ``stop_keyword``, ``manual``,
        ``complaint`` or ``invalid_number``). Never retried."""
        return self._t.request(
            "POST", "/v1/compliance/suppressions", json_body={"e164": e164, "reason": reason}
        )

    def import_(self, items: Iterable[Mapping[str, str]]) -> SuppressionImportResult:
        """Bulk import ``[{"e164", "reason"}]``. Named ``import_`` because
        ``import`` is a Python keyword; ``getattr(s, "import")`` also works.
        Never retried."""
        body = {"items": [{"e164": i["e164"], "reason": i["reason"]} for i in items]}
        return self._t.request("POST", "/v1/compliance/suppressions/import", json_body=body)

    def delete(self, id: Union[int, str]) -> None:
        """Remove a suppression."""
        self._t.request("DELETE", f"/v1/compliance/suppressions/{path_id(id)}")


# Expose the canonical SURFACE.md name too, reachable via getattr.
setattr(Suppressions, "import", Suppressions.import_)
